### PR TITLE
WIP: pm_layered for cpu/atmega_common

### PR DESCRIPTION
**Pull request by roberthartung**
This is a WIP PR for pm_layered in atmega_common. This has to be discussed on several parts, e.g.:

- How to actually disable/enable correct hardware devices?

### CODE REVIEW

File: cpu/atmega_common/periph/adc.c [hunk 1]
```diff
@@ -33,6 +36,10 @@ static mutex_t lock = MUTEX_INIT;
 static inline void _prep(void)
 {
     mutex_lock(&lock);
+#ifdef MODULE_PM_LAYERED
+    pm_block(PM_INVALID_ADC);
```
Comment (roberthartung) [inline cpu/atmega_common/periph/adc.c]: The define PM_INVALID_ADC defines the mode which the ADC needs to be running.

File: cpu/atmega_common/periph/spi.c [hunk 2]
```diff
@@ -38,11 +40,35 @@
 
 static mutex_t lock = MUTEX_INIT;
 
+
+static inline void spi_poweron(spi_t bus) {
+#ifdef MODULE_PM_LAYERED
+    pm_block(PM_INVALID_SPI);
+#endif
+    switch(bus) {
```
Comment (roberthartung) [inline cpu/atmega_common/periph/spi.c]: Is there a better way than a switch/case?

File: cpu/atmega_common/periph/spi.c [hunk 3]
```diff
@@ -66,11 +92,13 @@ int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
...
     /* lock the bus and power on the SPI peripheral */
     mutex_lock(&lock);
-    power_spi_enable();
+    /* power on and block sleep level */
+    spi_poweron(bus);
 
     /* configure as master, with given mode and clock */
     SPSR = (clk >> S2X_SHIFT);
     SPCR = ((1 << SPE) | (1 << MSTR) | mode | (clk & CLK_MASK));
+    //SPCR |= (1 << SPE);
```
Comment (roberthartung) [inline cpu/atmega_common/periph/spi.c]: Duplicate line... right?

File: cpu/atmega_common/periph/timer.c [hunk 4]
```diff
@@ -58,6 +62,7 @@ typedef struct {
     void *arg;                  /**< interrupt callback argument */
     uint8_t mode;               /**< remember the configured mode */
     uint8_t isrs;               /**< remember the interrupt state */
+    int8_t num;                /** < hardware timer number */
```
Comment (roberthartung) [inline cpu/atmega_common/periph/timer.c]: This is an approach where I map the timer within RIOT to the actual hardware timer. Any better solutions?

File: cpu/atmega_common/periph/timer.c [hunk 5]
```diff
@@ -85,6 +90,113 @@ static ctx_t *ctx[] = {{ NULL }};
 #endif
 /** @} */
 
+static inline void timer_poweroff(tim_t tim) {
+    switch(ctx[tim].num) {
```
Comment (roberthartung) [inline cpu/atmega_common/periph/timer.c]: Again, big switch/case

File: cpu/atmega_common/periph/timer.c [hunk 6]
```diff
@@ -93,6 +205,42 @@ int timer_init(tim_t tim, unsigned long freq, timer_cb_t cb, void *arg)
     DEBUG("timer.c: freq = %ld\n", freq);
     uint8_t pre = 0;
 
+
+    #ifdef MEGA_TIMER0
+    if(ctx[tim].dev == MEGA_TIMER0) {
+        ctx[tim].num = 0;
```
Comment (roberthartung) [inline cpu/atmega_common/periph/timer.c]: Maybe there is a better/more elegant way of setting the hardware timer?

File: cpu/atmega_common/periph/uart.c [hunk 7]
```diff
@@ -123,44 +143,122 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+    isr_ctx[num].rx_cb(isr_ctx[num].arg, ctx[num].dev->DR);
 
     if (sched_context_switch_request) {
         thread_yield();
     }
 }
 
+void uart_poweron(uart_t uart) {
+    switch(ctx[uart].num) {
+#ifdef MEGA_UART0
```
Comment (roberthartung) [inline cpu/atmega_common/periph/uart.c]: Do we need a differentiation if we have rx/tx enabled or not? See discussion in #7947

**Comment by miri64** [comment]
I'm not sure why I self-assigned for this... Maybe @MichelRottleuthner @kaspar030 or @haukepetersen want to take a look instead?

**Comment by roberthartung** [comment]
@miri64 Can you remove the label? Now that #11122 is merged, I can finally work on the pm_layered module again (can finally test it with our platform that needs #11122).

**Comment by miri64** [comment]
> @miri64 Can you remove the label?

Stale bot did it already in reaction to your comment ;-)

**Comment by maribu** [comment]
@roberthartung: Feel free to ping me when this PR has been rebased and got into a state a review makes sense. I really would like to see the power management of the ATmega platform improved in RIOT.

**Comment by roberthartung** [comment]
@maribu Thanks! Will work on this as one of the next steps. Are there enough tests for pm already? I should also add my pm test to this PR:
- A simple test that consists of an RTT and enters a low-power sleep mode and then wakes up from it using the RTT.

**Comment by miri64** [comment]
@MrKevinWeiss can probably answer that.

**Comment by MrKevinWeiss** [comment]
I haven't touch pm stuff yet in the testing.  I am working on a way to do so with the HiL setup but I don't think it would be ready in time.

I can however run some tests manually and can also monitor the current (I think).

I think having a simple (on target) test that turns things off and on is great.  You can also try failure cases where you try to turn the power off when something is held on.  Or what happens when turning power on twice...

That kind of thing.

The easier the test is to run the easier it is to merge in.

My "helping people" time is currently with the release and esp8266 PR but after I clear those away I would be happy work on how to test this.

**Comment by benpicco** [comment]
> A simple test that consists of an RTT and enters a low-power sleep mode and then wakes up from it using the RTT.

`tests/periph_pm` will do just that.
Unfortunately it uses the RTC instead of RTT, but that should be an easy change.

**Pull request closed without merging**